components/feature/session/src/main/java/mozilla/components/feature/session/ThumbnailsFeature.kt

```diff
@@ -0,0 +1,69 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.session
+
+import android.content.Context
+import android.support.annotation.VisibleForTesting
+import mozilla.components.browser.session.SelectionAwareSessionObserver
+import mozilla.components.browser.session.Session
+import mozilla.components.browser.session.SessionManager
+import mozilla.components.concept.engine.EngineView
+import mozilla.components.support.base.feature.LifecycleAwareFeature
+import mozilla.components.support.ktx.android.content.isOSOnLowMemory
+
+/**
+ * Feature implementation for automatically taking thumbnails of sites.
+ */
+class ThumbnailsFeature(
+    private val context: Context,
+    private val engineView: EngineView,
+    sessionManager: SessionManager
+) : LifecycleAwareFeature {
+
+    private val observer = ThumbnailsFeatureRequestObserver(sessionManager)
+
+    /**
+     * Starts observing the selected session to listen for when a session finish loading.
+     */
+    override fun start() {
+        observer.observeSelected()
+    }
+
+    /**
+     * Stops observing the selected session.
+     */
+    override fun stop() {
+        observer.stop()
+    }
+
+    internal inner class ThumbnailsFeatureRequestObserver(
+        sessionManager: SessionManager
+    ) : SelectionAwareSessionObserver(sessionManager) {
+
+        override fun onProgress(session: Session, progress: Int) {
+            if (progress == PROGRESS_STOP) {
+                requestPrintScreen(session)
+            }
+        }
+    }
+
+    private fun requestPrintScreen(session: Session) {
+        if (!isLowOnMemory()) {
+            engineView.captureThumbnail {
+                session.thumbnail = it
+            }
+        }
```

Comment:
nit: I assume in a low memory situation we would want to assign `null`? Otherwise we will have a "ghost" image in low memory situations that will never update/change.

Reply:
When you said "ghost image", did you refer to the previous  `session.thumbnail`?
Because, right now if we are under low memory conditions, we are not requesting a new thumbnail.

```kotlin
        if (!isLowOnMemory()) {
            engineView.captureThumbnail {
                session.thumbnail = it
            }
        }
```

Did you mean `session.thumbnail  = null` on  low memory?

Reply:
Yep. Otherwise this will keep showing the old image. Better no image than an outdated/wrong one?

Reply:
Yeah, I didn't add it initially because we are already doing it on `SessionManager`. 

  ```kotlin
  /**
     * Informs this [SessionManager] that the OS is in low memory condition so it
     * can reduce its allocated objects.
     */
    fun onLowMemory() {
        // Removing the all the thumbnails except for the selected session to
        // reduce memory consumption.
        sessions.forEach {
            if (it != selectedSession) {
                it.thumbnail = null
            }
        }
}
```

https://github.com/mozilla-mobile/android-components/blob/024e3de456e3b46e9bf6718db9500ecc52da3d29/components/browser/session/src/main/java/mozilla/components/browser/session/SessionManager.kt#L472